### PR TITLE
Fix parsing AM/PM hours

### LINKED ISSUE #9333: Time.parse doesn't work for 12 hour times between noon and 1pm
I may be crazy, but this:

`puts Time.parse("2020-10-10 12:30 PM", "%Y-%m-%d %l:%M %P", Time::Location::UTC).inspect`

Throws:

```Unhandled exception: Invalid time (ArgumentError)
  from /usr/local/Cellar/crystal/0.34.0/src/time.cr:417:7 in 'local:nanosecond:location'
  from /usr/local/Cellar/crystal/0.34.0/src/time/format/parser.cr:55:16 in 'time'
  from /usr/local/Cellar/crystal/0.34.0/src/time/format.cr:84:5 in 'parse'
  from /usr/local/Cellar/crystal/0.34.0/src/time/format.cr:81:3 in 'parse'
  from /usr/local/Cellar/crystal/0.34.0/src/time.cr:1163:5 in 'parse'
```

However, other 12 hour times work just fine:

`puts Time.parse("2020-10-10 01:30 PM", "%Y-%m-%d %l:%M %P", Time::Location::UTC).inspect`

Am I missing something? The mention of `local:nanosecond:location` makes me think it might be a bug?  Thanks for an awesome language!

### PR DESCRIPTION
**Pull request by straight-shoota**
Fixes #9333

### CODE REVIEW

**Comment by jhass** [comment]
Not quite there yet. 

```
bin/crystal eval 'p Time.parse("12", "%H", Time::Location::UTC)'
Using compiled compiler at .build/crystal
0001-01-01 00:00:00.0 UTC
```

I think we need to turn `@pm` into a tri-state of "AM", "PM", "24hour clock"

**Comment by straight-shoota** [comment]
So we're missing even more specs... =)

**Comment by jhass** [comment]
Might also want to add some to verify `%l` on its own without `%P` parses to am.

File: src/time/format/parser.cr [hunk 1]
```diff
@@ -32,12 +32,21 @@ struct Time::Format
...
-      @hour += 12 if @pm
+      if @pm
+        unless @hour == 12
+          @hour += 12
+        end
+      elsif @am
+        if @hour == 12
+          @hour = 0
+        end
+      end
```
Comment (jhass) [inline src/time/format/parser.cr:49]: How about folding the nested conditions?

```suggestion
      if @pm && @hour != 12
        @hour += 12
      elsif @am && @hour == 12
        @hour = 0
      end
```

**Comment by straight-shoota** [comment]
I found another issue and fixed it in 9a90d33: On the 24-hour clock, 24:00 is a valid time indicating midnight and equivalent to 00:00 of the next day (see ISO 8601 §5.3). The validation in `Time#initializer` incorrectly rejected this.

**Comment by jhass** [comment]
```
bin/crystal eval 'p Time.parse("12", "%l", Time::Location::UTC)'
Using compiled compiler at .build/crystal
0001-01-01 12:00:00.0 UTC
```

Should this be `00:00`?

**Comment by straight-shoota** [comment]
Hm, I guess either could be right?

Ruby parses `12` with `%I` as `12`. libc's `strptime` parses it as `0`.

Maybe Ruby's implementation didn't think about this case. And it's really rare.
I have not preference which one it should be.

**Comment by jhass** [comment]
Well, for 0-11 it parses as AM, so why not parse the full range as AM?

**Comment by straight-shoota** [comment]
Okay, I'll use that. That means `Time.parse("00", "%I")` should be invalid, at least it's in both Python and Ruby (which seems to indicate that Ruby's behaviour might not be fully intentional).

**Comment by straight-shoota** [comment]
Also added a spec for when 24-hour clock hour is used with am/pm (which is ignored).

**Comment by jhass** [comment]
> Okay, I'll use that. That means Time.parse("00", "%I") should be invalid, at least it's in both Python and Ruby (which seems to indicate that Ruby's behaviour might not be fully intentional).

So right now this rejects `0` for `%l` but accepts  `0am` for `%l%P`, but the rationale was that if no `%P` is given for `%l`/`%I`, then we assume AM. And both Ruby and Python seem to follow that rationale and reject `0AM` generally.

I don't care too much which way we go, looking at https://en.wikipedia.org/wiki/12-hour_clock#Confusion_at_noon_and_midnight `0AM` seems to be used by the Japanese at least sometimes, so I'd slightly favor allowing it. But `0` and `0AM` should parse the same in all cases for `%l` and `%l%P` respectively.

**Comment by straight-shoota** [comment]
Oh yeah `0 am` should be invalid, too. I'm not sure why I added it...

File: src/time/format/parser.cr [hunk 2]
```diff
@@ -32,12 +32,28 @@ struct Time::Format
...
+        else
+          if @hour == 0
+            raise ArgumentError.new("Invalid hour for 12-hour clock")
+          end
+
+          if @hour == 12
+            @hour = 0
+          end
+        end
+      end
```
Comment (jhass) [inline src/time/format/parser.cr:56]: `@am` is no longer used.

Comment (jhass) [inline src/time/format/parser.cr:56]: YMMV but for me postfix if's make this a little easier to read and follow.
```suggestion
      if @hour_is_12
        if @pm
          @hour += 12 if @hour != 12
        else # AM
          raise ArgumentError.new("Invalid hour for 12-hour clock") if @hour == 0

          @hour = 0 if @hour == 12
        end
      end
```

**Pull request merged**